### PR TITLE
Correctly lock calls to removeIntermediateSolution.

### PR DESCRIPTION
**Pull request by lieblb**
In the test module, some UI calls to `removeIntermediateSolution` are sometimes not correctly locked using the test module's process locker strategy, but are instead performed without locking.

For example, look at this stack trace from ILIAS 5.3.10:

```
#12 ilDBPdo:manipulate in /var/www/html/ILIAS/Services/Database/classes/PDO/class.ilDBPdo.php:1044
#11 ilDBPdo:manipulateF in /var/www/html/ILIAS/Modules/TestQuestionPool/classes/class.assQuestion.php:4961
#10 assQuestion:removeCurrentSolution in /var/www/html/ILIAS/Modules/TestQuestionPool/classes/class.assQuestion.php:4920
#9 assQuestion:removeIntermediateSolution in /var/www/html/ILIAS/Modules/Test/classes/class.ilTestOutputGUI.php:447
#8 ilTestOutputGUI:submitSolutionCmd in /var/www/html/ILIAS/Modules/Test/classes/class.ilTestOutputGUI.php:143
#7 ilTestOutputGUI:executeCommand in /var/www/html/ILIAS/Services/UICore/classes/class.ilCtrl.php:197
#6 ilCtrl:forwardCommand in /var/www/html/ILIAS/Modules/Test/classes/class.ilObjTestGUI.php:212
#5 ilObjTestGUI:executeCommand in /var/www/html/ILIAS/Services/UICore/classes/class.ilCtrl.php:197
#4 ilCtrl:forwardCommand in /var/www/html/ILIAS/Services/Repository/classes/class.ilRepositoryGUI.php:402
#3 ilRepositoryGUI:show in /var/www/html/ILIAS/Services/Repository/classes/class.ilRepositoryGUI.php:356
#2 ilRepositoryGUI:executeCommand in /var/www/html/ILIAS/Services/UICore/classes/class.ilCtrl.php:197
#1 ilCtrl:forwardCommand in /var/www/html/ILIAS/Services/UICore/classes/class.ilCtrl.php:159
#0 ilCtrl:callBaseClass in /var/www/html/ILIAS/ilias.php:21
```

In some configurations (like enabling full transactions for the test module, which I'm currently working on), this will lead to various database exceptions. In standard ILIAS installations, this just allows operations on the test tables to run without a lock, even though a lock is configured.

This PR adds a lock to `ilTestPlayerAbstractGUI.removeIntermediateSolution`.

I tested this change using TestILIAS with the standard "use database locks" setting in the test settings and didn't experience any problems (e.g. deadlocks).

### CODE REVIEW

**Comment by bheyser** [comment]
The test & assessment lock strategy does avoid race conditions in general.

For which race condition this lock is required? Behind removeIntermediateSolution there is a single delete statement that imho cannot lead to any race condition with duplicate requests.

**Comment by mjansenDatabay** [comment]
@lieblb Could you please provide a feedback for @bheyser ?

**Comment by lieblb** [comment]
I talked about this with Björn on the phone a while back I think. Let me summarize. In short, I believe it should be addressed in some way.

### Vanilla ILIAS

Assume some Process A is inside a call to `assMultipleChoice::saveWorkingData` with `$authorized == false` (i.e. in the process of saving a multiple choice intermediate solution):

https://github.com/ILIAS-eLearning/ILIAS/blob/d7a8c6af29c995534f0f67be8e11a56ff24c57d2/Modules/TestQuestionPool/classes/class.assMultipleChoice.php#L672-L685

After the process obtained a lock via `executeUserSolutionUpdateLockOperation`, the process deletes all intermediate solutions with `removeCurrentSolution` and then puts in each checked value from the MC question with `saveCurrentSolution`.


Another Process B for the same user has just saved an answer (used a properly locked call to `saveQuestionSolution`) and is now inside the unlocked part of `submitSolutionAndNextCmd` where it is about to remove the user's intermediate solutions via `removeIntermediateSolution`:

https://github.com/ILIAS-eLearning/ILIAS/blob/d7a8c6af29c995534f0f67be8e11a56ff24c57d2/Modules/Test/classes/class.ilTestOutputGUI.php#L411-L419


I believe you have a race condition here: process A might have saved one of two check box values (say `X` of `[X, Y]`) through `saveCurrentSolution`, when suddenly, without a lock, Process B's `removeIntermediateSolution` interferes and removes all saved values. Process A then continues to save `Y` into the now cleaned up table. You end up with a wrong intermediate answer in the DB that lacks the first few values.

### TA Transaction Patches

I originally came across these calls to `removeIntermediateSolution` as issues when working on full DB transactions for the TA module; see https://gitlab.com/lieblb/ilias/tree/edge/ur.5.3/Customizing/uni-regensburg/extensions/Transactions. There, DB transactions are started and committed through the existing TA module's `getProcessLocker` mechanism, see https://gitlab.com/lieblb/ilias/blob/edge/ur.5.3/Customizing/uni-regensburg/extensions/Transactions/classes/class.urAssQuestionProcessLockerDb.php and https://gitlab.com/lieblb/ilias/blob/edge/ur.5.3/Customizing/uni-regensburg/extensions/Transactions/classes/class.urNestableAtomQueryTransaction.php.

In short, without this PR's patch and the transaction logic just described in place, MySQL would throw errors and abort transaction due to unlocked calls to `removeIntermediateSolution`. One process modified `tst_solutions` through a proper transaction/lock, while another called `removeIntermediateSolution` which did not obtain a lock/transaction - the latter then aborted with an obscure MySQL error. I have to admit I can no longer reproduce this behaviour (I would have provided error logs otherwise).
